pkg/kubelet/kuberuntime/logs/logs_test.go

```diff
@@ -454,3 +453,67 @@ func TestReadLogsLimitsWithTimestamps(t *testing.T) {
 
 	assert.Equal(t, 2, lineCount, "should have two lines")
 }
+
+func TestReadLogsWithLogRotation(t *testing.T) {
+	logLineFmt := "2023-06-09T05:28:57.806368292Z stdout F %v\n"
+
+	dir, err := os.MkdirTemp("", "readlogs")
+	assert.NoError(t, err)
+	defer os.RemoveAll(dir)
+	tmpfile, err := os.CreateTemp(dir, "log.*.txt")
+	assert.NoError(t, err)
+
+	for i := 0; i < 5; i++ {
+		tmpfile.WriteString(fmt.Sprintf(logLineFmt, i))
+	}
+	tmpfile.Close()
+
+	ctx := context.Background()
+	// use fake runtime service
+	containerID := "83a21271e2a84a48ac89-6411e55fb8c1"
+	fakeRuntimeService := apitesting.NewFakeRuntimeService()
+	fakeRuntimeService.Containers[containerID] = &apitesting.FakeContainer{
+		ContainerStatus: runtimeapi.ContainerStatus{
+			Id:    containerID,
+			State: runtimeapi.ContainerState_CONTAINER_RUNNING,
+		},
+	}
+
+	// mock log rotation
+	go func() {
+		time.Sleep(1 * time.Second)
+
+		os.Rename(tmpfile.Name(), tmpfile.Name()+".1")
+		defer os.Remove(tmpfile.Name() + ".1")
+
+		// call CRI.ReopenContainerLog took some time
+		time.Sleep(500 * time.Millisecond)
+		newTmpfile, err := os.OpenFile(tmpfile.Name(), os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
+		assert.NoError(t, err)
+
+		for i := 5; i < 10; i++ {
+			newTmpfile.WriteString(fmt.Sprintf(logLineFmt, i))
+		}
+		newTmpfile.Close()
+
+		// wait 1 second and stop container
+		time.Sleep(1 * time.Second)
+		fakeRuntimeService.Containers[containerID].ContainerStatus.State = runtimeapi.ContainerState_CONTAINER_EXITED
+	}()
+
+	var buf bytes.Buffer
+	w := io.MultiWriter(&buf)
+
+	err = ReadLogs(ctx, tmpfile.Name(), containerID,
+		&LogOptions{tail: -1, bytes: -1, timestamp: true, follow: true},
+		fakeRuntimeService, w, w)
+	assert.NoError(t, err)
+
+	lineCount := 0
+	scanner := bufio.NewScanner(bytes.NewReader(buf.Bytes()))
+	for scanner.Scan() {
+		lineCount++
+	}
+
+	assert.Equal(t, 10, lineCount, "should have 10 lines")
+}
```

Comment:
Would it make sense to modify `TestReadLogs` instead of writing new test?

Reply:
The handling of this test differs slightly from the case in TestReadLogs.

Reply:
If it differs slightly it's a good chance to modify it. It looks a bit strange to have to tests testing the same function.

Reply:
`TestReadLogs` currently includes some test cases specific to `PodLogOptions`. I suggest separating the tests for different scenarios, since we already have other tests that cover these scenarios independently.

```
TestLogOptions
TestReadLogs
TestParseLog
TestWriteLogs
TestWriteLogsWithBytesLimit
TestReadLogsLimitsWithTimestamps
TestReadLogsWithLogRotation
```

Reply:
makes sense to me, thanks.